api/topo/topo.proto

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2019-present Open Networking Foundation.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+syntax = "proto3";
+
+// Package topo defines interfaces for managing network top
+package topo;
+
+import "google/protobuf/any.proto";
+import "gogoproto/gogo.proto";
+
+// EntityService provides an API for managing entities.
+service topo {
+  // Update one or more entities to the topology
+  rpc Write (WriteRequest) returns (WriteResponse) {
+  }
+
+  // Read one or more entities from topology
+  rpc Read (ReadRequest) returns (ReadResponse) {
+  }
+
+  // Represents the bidirectional stream between
+  // onos-topo and a client for the purpose of
+  // - streaming notifications
+  rpc StreamChannel (stream StreamMessageRequest) returns (stream StreamMessageResponse) {
+  }
+}
+
+message WriteRequest {
+  // The write batch, comprising a list of Update operations
+  repeated Update updates = 1;
+}
+
+message WriteResponse {
+}
+
+message ReadRequest {
+  repeated Object objects = 1;
+}
+
+message ReadResponse {
+  repeated Object objects = 1;
+}
+
+message StreamMessageRequest {
+}
+
+message StreamMessageResponse {
+}
+
+message Update {
+  enum Type {
+    UNSPECIFIED_TYPE = 0;
+    INSERT = 1;
+    MODIFY = 2;
+    DELETE = 3;
+  }
+  Type type = 2;
+
+  Object object = 3;
+}
+
+message Object {
+  string id = 1 [(gogoproto.customname) = "ID", (gogoproto.casttype) = "ID"];
+
+  enum Type {
+    UNSPECIFIED_TYPE = 0;
+    ENTITY = 1;
+    RELATIONSHIP = 2;
+  }
+  Type type = 2;
+
+  oneof obj {
+    Entity entity = 3;
+    Relationship relationship = 4;
+  }
+}
+
+// Entity represents any "thing" that is represented in the topology
+message Entity {
+  string type = 1;
+
+  // attributes is an arbitrary mapping of attribute keys (strings) to values (any)
+  map<string, google.protobuf.Any> attributes = 2;
+
+}
+
+message Relationship {
```

Comment:
I think that relationships should also have attributes so I think we may want to re-structure things slightly, but otherwise this is on the right track. Let me think about this... I'll provide a concrete suggestion for restructuring.

Reply:
done